src/tools/rbd_mirror/InstanceReplayer.cc

```diff
@@ -0,0 +1,494 @@
+          dout(20) << global_image_id << ": unknown local_image_id"
+                   << " (image not exist or primary), skipping delete" << dendl;
+        } else {
+          m_image_deleter->schedule_image_delete(
+            m_local_rados, m_local_pool_id, local_image_id, global_image_id);
+        }
+        on_finish->complete(0);
+      });
+  }
+
+  stop_image_replayer(image_replayer, on_finish);
+}
+
+template <typename I>
+void InstanceReplayer<I>::print_status(Formatter *f, stringstream *ss) {
+  dout(20) << dendl;
+
+  if (!f) {
+    return;
+  }
+
+  Mutex::Locker locker(m_lock);
+
+  f->open_array_section("image_replayers");
+  for (auto &kv : m_image_replayers) {
+    auto &image_replayer = kv.second;
+    image_replayer->print_status(f, ss);
+  }
+  f->close_section();
+}
+
+template <typename I>
+void InstanceReplayer<I>::start()
+{
+  dout(20) << dendl;
+
+  Mutex::Locker locker(m_lock);
+
+  m_manual_stop = false;
+
+  for (auto &kv : m_image_replayers) {
+    auto &image_replayer = kv.second;
+    image_replayer->start(nullptr, true);
+  }
+}
+
+template <typename I>
+void InstanceReplayer<I>::stop()
+{
+  dout(20) << dendl;
+
+  Mutex::Locker locker(m_lock);
+
+  m_manual_stop = true;
+
+  for (auto &kv : m_image_replayers) {
+    auto &image_replayer = kv.second;
+    image_replayer->stop(nullptr, true);
+  }
+}
+
+template <typename I>
+void InstanceReplayer<I>::restart()
+{
+  dout(20) << dendl;
+
+  Mutex::Locker locker(m_lock);
+
+  m_manual_stop = false;
+
+  for (auto &kv : m_image_replayers) {
+    auto &image_replayer = kv.second;
+    image_replayer->restart();
+  }
+}
+
+template <typename I>
+void InstanceReplayer<I>::flush()
+{
+  dout(20) << "enter" << dendl;
+
+  Mutex::Locker locker(m_lock);
+
+  for (auto &kv : m_image_replayers) {
+    auto &image_replayer = kv.second;
+    image_replayer->flush();
+  }
+}
+
+template <typename I>
+void InstanceReplayer<I>::start_image_replayer(ImageReplayer<I> *image_replayer) {
+  assert(m_lock.is_locked());
+
+  std::string global_image_id = image_replayer->get_global_image_id();
+  dout(20) << "global_image_id=" << global_image_id << dendl;
+
+  if (!image_replayer->is_stopped()) {
+    return;
+  } else if (image_replayer->is_blacklisted()) {
+    derr << "blacklisted detected during image replay" << dendl;
+    return;
+  }
+
+  FunctionContext *ctx = new FunctionContext(
+    [this, global_image_id] (int r) {
+      dout(20) << "image deleter result: r=" << r << ", "
+               << "global_image_id=" << global_image_id << dendl;
+      if (r == -ESTALE || r == -ECANCELED) {
+        return;
+      }
+
+      Mutex::Locker locker(m_lock);
+      auto it = m_image_replayers.find(global_image_id);
+      if (it == m_image_replayers.end()) {
+        return;
+      }
+
+      auto image_replayer = it->second;
+      if (r >= 0) {
+        image_replayer->start(nullptr, false);
+      } else {
+        start_image_replayer(image_replayer);
+      }
+    });
+
```

Comment:
I think we need to start and op here and finish it in the context above before each return point since we need to ensure zero percent possibility of the ```ImageDeleter``` calling back into a deleted instance replayer.